packages/mongodb-chatbot-server/src/routes/generateResponse.ts

```diff
@@ -35,61 +42,147 @@ interface GenerateResponseReturnValue {
   messages: SomeMessage[];
 }
 
+export type GenerateResponse = (
+  params: GenerateResponseParams
+) => Promise<GenerateResponseReturnValue>;
+
 /**
   Generate a response with/without streaming. Supports tool calling
   and standard response generation.
+  Response includes the user message with any data mutations
+  and the assistant response message, plus any intermediate tool calls.
  */
 export async function generateResponse({
   shouldStream,
   llm,
-  llmConversation,
+  latestMessageText,
+  customData,
+  generateUserPrompt,
+  filterPreviousMessages,
   dataStreamer,
-  references,
   reqId,
   llmNotWorkingMessage,
   noRelevantContentMessage,
   conversation,
   request,
 }: GenerateResponseParams): Promise<GenerateResponseReturnValue> {
+  const { userMessage, references, staticResponse, rejectQuery } =
+    await (generateUserPrompt
+      ? generateUserPrompt({
+          userMessageText: latestMessageText,
+          conversation,
+          reqId,
+          customData,
+        })
+      : {
+          userMessage: {
+            role: "user",
+            content: latestMessageText,
+            customData,
+          } satisfies UserMessage,
+        });
+  // Add request custom data to user message.
+  const userMessageWithCustomData = customData
+    ? {
+        ...userMessage,
+        // Override request custom data fields with user message custom data fields.
+        customData: { ...customData, ...(userMessage.customData ?? {}) },
+      }
+    : userMessage;
+  const newMessages: SomeMessage[] = [userMessageWithCustomData];
+
+  // Send static response if query rejected or static response provided
+  if (rejectQuery) {
+    const rejectionMessage = {
+      role: "assistant",
+      content: noRelevantContentMessage,
+      references: references ?? [],
+    } satisfies AssistantMessage;
+    newMessages.push(rejectionMessage);
+  } else if (staticResponse) {
+    newMessages.push(staticResponse);
+  }
+
+  // Prepare conversation messages for LLM
+  const previousConversationMessagesForLlm = (
+    filterPreviousMessages
+      ? await filterPreviousMessages(conversation)
+      : conversation.messages
+  ).map(convertConversationMessageToLlmMessage);
+  const newMessagesForLlm = newMessages.map((m) => {
+    // Use transformed content if it exists for user message,
```

Comment:
```suggestion
    // Use transformed content (e.g. from a custom user prompt, query preprocessor, etc) if it exists for user message,
```

Small change for clarity to this old comment